### Домашнее задание 5 - 10 баллов

В этом задании вам предстоит дообучить трансформерную модель для задачи классификации с помощью различных техник и сравнить их между собой.

Датасет: [dair-ai/emotion](https://huggingface.co/datasets/dair-ai/emotion) 

Модель: [google-bert/bert-base-uncased](https://huggingface.co/google-bert/bert-base-uncased) (если хочется, можно заменить на что-то более интересное)

1. Скачайте датасет и модель. Измерьте базовые метрики классификации перед началом экспериментов.

**NB!** Для всех типов дообучения замерьте :
- качество классификации на выходе
- время дообучения
- количество параметров для обучения
- потребление ресурсов (не нужно заморачиваться с профайлингом - можно просто посмотреть в `nvidia-smi` или `torch.cuda.memory_allocated`)

2. Обучите модель в режиме full finetuning - **1 балл**
3. Обучите модель в режиме linear probing - реализуйте кастомную классификационную голову и обучайте только ее. Не забудьте описать, чем обусловлено устройство головы, как вы пришли к такой архитектуре - **2 балла**
4. Обучите модель в режиме PEFT с использованием [prompt tuning или prefix tuning](https://ericwiener.github.io/ai-notes/AI-Notes/Large-Language-Models/Prompt-Tuning-and-Prefix-Tuning). При выборе метода напишите пару слов, почему решили остановиться именно на этом методе - **2 балла**
4. Обучите модель в режиме PEFT с использованием LoRA. Попробуйте подобрать оптимальный ранг - `r`, при желании поэкспериментируйте с остальными гиперпараметрами. Опишите, чем обусловлена ваша финальная конфигурация - **2 балла**

5. Соберите все результаты отдельных замеров в таблицу и сделайте выводы о вычислительной сложности методов, итоговом качестве и прочих наблюдаемых свойствах моделей - **1 балл**

**Общее**

- Принимаемые решения обоснованы (почему выбрана определенная архитектура/гиперпараметр/оптимизатор/преобразование и т.п.) - **1 балл**
- Обеспечена воспроизводимость решения: зафиксированы random_state, ноутбук воспроизводится от начала до конца без ошибок - **1 балл**

In [ ]:
import torch
import numpy as np
import random
import evaluate
import time
import pandas as pd

from datasets import load_dataset
from transformers import AutoTokenizer, AutoModelForSequenceClassification, TrainingArguments, Trainer
from peft import get_peft_model, LoraConfig, PrefixTuningConfig, TaskType, PromptTuningConfig
from transformers import DataCollatorWithPadding

SEED = 42
random.seed(SEED)
np.random.seed(SEED)
torch.manual_seed(SEED)
torch.cuda.manual_seed_all(SEED)

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# 1. Загрузка датасета и модели

In [2]:
dataset = load_dataset("dair-ai/emotion")
label_list = dataset["train"].features["label"].names
num_labels = len(label_list)

model_name = "bert-base-uncased"
tokenizer = AutoTokenizer.from_pretrained(model_name)


def tokenize(example):
    return tokenizer(example["text"], truncation=True)


dataset = dataset.map(tokenize, batched=True)
data_collator = DataCollatorWithPadding(tokenizer=tokenizer)

In [3]:
pd.Series(dataset["train"]["label"]).value_counts()

1    5362
0    4666
3    2159
4    1937
2    1304
5     572
Name: count, dtype: int64

In [ ]:
# DISCLAIMER: для чистоты эксперимента и из-за ограничения по времени, параметры обучения типа
# lr, количества эпох и тд. подбираться не будут, чтобы посмотреть сходимость за фиксированное количество эпох
# + для оценки метрик в конце используется лучшая модель среди всех эпох по минимальному валидационному лоссу
args = TrainingArguments(
    output_dir="./hw_5",
    eval_strategy="epoch",
    save_strategy="epoch",
    per_device_train_batch_size=16,
    per_device_eval_batch_size=64,
    num_train_epochs=10,
    learning_rate=1e-4,
    seed=SEED,
    load_best_model_at_end=True,
)

In [5]:
model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=num_labels).to(device)

accuracy_metric = evaluate.load("accuracy")
recall_metric = evaluate.load("recall")
precision_metric = evaluate.load("precision")
f1_metric = evaluate.load("f1")


def compute_metrics(eval_pred):
    logits, labels = eval_pred
    predictions = np.argmax(logits, axis=-1)
    return {
        "Accuracy": accuracy_metric.compute(predictions=predictions, references=labels)["accuracy"],
        "Recall": recall_metric.compute(predictions=predictions, references=labels, average="macro", zero_division=0)[
            "recall"
        ],
        "Precision": precision_metric.compute(
            predictions=predictions, references=labels, average="macro", zero_division=0
        )["precision"],
        "F1": f1_metric.compute(predictions=predictions, references=labels, average="macro")["f1"],
    }


trainer = Trainer(
    model=model,
    args=args,
    train_dataset=dataset["train"],
    eval_dataset=dataset["validation"],
    processing_class=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics,
)

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [9]:
eval_results = trainer.evaluate(dataset["test"])
for metric, value in eval_results.items():
    print(f"{metric}: {value:.4f}")
print(f"Trainable parameters: {(sum(p.numel() for p in model.parameters() if p.requires_grad) / 1e6):.2f}, M")
print(f"CUDA memory used: {(torch.cuda.memory_allocated() / 1e6):.2f}MB")

eval_loss: 1.1865
eval_model_preparation_time: 0.0018
eval_Accuracy: 0.5775
eval_Recall: 0.3039
eval_Precision: 0.2751
eval_F1: 0.2425
Trainable parameters: 109.49, M
CUDA memory used: 1334.25MB


In [11]:
def train_and_pretty_print_results(trainer):
    start = time.time()
    trainer.train()
    end = time.time()

    eval_results = trainer.evaluate(dataset["test"])
    for metric, value in eval_results.items():
        print(f"{metric}: {value:.4f}")
    print("Training time:", round(end - start, 2), "sec")
    print(f"Trainable parameters: {(sum(p.numel() for p in trainer.model.parameters() if p.requires_grad) / 1e6):.2f}M")
    print(f"CUDA memory used: {(torch.cuda.memory_allocated() / 1e6):.2f}MB")

# 2. Full finetuning

In [12]:
train_and_pretty_print_results(trainer)

eval_loss: 0.1965
eval_model_preparation_time: 0.0018
eval_Accuracy: 0.9235
eval_Recall: 0.8891
eval_Precision: 0.8951
eval_F1: 0.8823
eval_runtime: 3.2845
eval_samples_per_second: 608.9210
eval_steps_per_second: 9.7430
epoch: 10.0000
Training time: 1127.64 sec
Trainable parameters: 109.49M
CUDA memory used: 1334.25MB


# 3. Linear probing

In [13]:
# загружаем модель и замораживаем encoder, обучается только голова
lp_model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=num_labels)
for param in lp_model.bert.parameters():
    param.requires_grad = False
lp_model = lp_model.to(device)

# используется просто CLS токен → Dropout → Dense → Softmax
# простой и эффективный классификатор, минимальные затраты, уже реализованный класс из transformers
# остальная часть кода аналогична Full finetuning, кроме `requires_grad = False`

trainer = Trainer(
    model=lp_model,
    args=args,
    train_dataset=dataset["train"],
    eval_dataset=dataset["validation"],
    processing_class=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics,
)
train_and_pretty_print_results(trainer)

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


eval_loss: 1.4338
eval_Accuracy: 0.4645
eval_Recall: 0.2388
eval_Precision: 0.1532
eval_F1: 0.1858
eval_runtime: 3.3332
eval_samples_per_second: 600.0290
eval_steps_per_second: 9.6000
epoch: 10.0000
Training time: 315.08 sec
Trainable parameters: 0.00M
CUDA memory used: 1774.06MB


# 4. Prompt/Prefix tuning

In [ ]:
# prefix tuning мне показался интереснее prompt tuning, потому что он добавляет обучаемые векторы прямо
# в attention внутри модели, а не просто в начало текста. то есть влияние этих токенов по идее должно быть
# глубже и сильнее, что может быть полезно в задачах с тонкими различиями, как эмоции. да, он чуть тяжелее
# по памяти, но не сильно. также прочитала, что он лучше работает на меньших моделях (типа bert)

peft_config = PrefixTuningConfig(
    task_type=TaskType.SEQ_CLS,
    num_virtual_tokens=20,  # средний стартовый вариант для большинства задач
)

peft_model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=num_labels)
peft_model = get_peft_model(peft_model, peft_config)
peft_model.print_trainable_parameters()
peft_model.to(device)

trainer = Trainer(
    model=peft_model,
    args=args,
    train_dataset=dataset["train"],
    eval_dataset=dataset["validation"],
    processing_class=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics,
)
train_and_pretty_print_results(trainer)

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


trainable params: 368,640 || all params: 109,855,494 || trainable%: 0.3356


No label_names provided for model class `PeftModelForSequenceClassification`. Since `PeftModel` hides base models input arguments, if label_names is not given, label_names can't be set automatically within `Trainer`. Note that empty label_names list will be used instead.


eval_loss: 1.5779
eval_Accuracy: 0.3820
eval_Recall: 0.1901
eval_Precision: 0.1284
eval_F1: 0.1400
eval_runtime: 3.3695
eval_samples_per_second: 593.5530
eval_steps_per_second: 9.4970
epoch: 10.0000
Training time: 581.81 sec
Trainable parameters: 0.37M
CUDA memory used: 2256.85MB


In [ ]:
# однако из-за того, что prefix tuning показал такие плохие метрики (как бы я не меняла параметры), решила попробовать и prompt tuning
# в итоге я остановилась на нем из-за лучших метрик. он позволяет дообучать только маленькое количество виртуальных токенов,
# которые добавляются в начало входа. это суперэффективный способ, особенно если нет возможности гонять всю модель.
# ещё один плюс — его легко внедрить через готовые инструменты, и он не требует изменений внутри самой модели.

peft_config = PromptTuningConfig(
    task_type=TaskType.SEQ_CLS,
    prompt_tuning_init="TEXT",
    prompt_tuning_init_text="Classify the emotion in this sentence.",
    num_virtual_tokens=20,  #
    tokenizer_name_or_path=model_name,
)

peft_model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=num_labels)
peft_model = get_peft_model(peft_model, peft_config)
peft_model.print_trainable_parameters()
peft_model.to(device)

trainer = Trainer(
    model=peft_model,
    args=args,
    train_dataset=dataset["train"],
    eval_dataset=dataset["validation"],
    processing_class=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics,
)
train_and_pretty_print_results(trainer)

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


trainable params: 15,360 || all params: 109,502,214 || trainable%: 0.0140


No label_names provided for model class `PeftModelForSequenceClassification`. Since `PeftModel` hides base models input arguments, if label_names is not given, label_names can't be set automatically within `Trainer`. Note that empty label_names list will be used instead.


eval_loss: 1.4769
eval_Accuracy: 0.4750
eval_Recall: 0.2989
eval_Precision: 0.2231
eval_F1: 0.2508
eval_runtime: 4.5304
eval_samples_per_second: 441.4670
eval_steps_per_second: 7.0630
epoch: 10.0000
Training time: 780.66 sec
Trainable parameters: 0.02M
CUDA memory used: 1804.14MB


# 5. LoRA

In [ ]:
for r in [2, 4, 8, 16, 32]:
    print("R = ", r)
    lora_config = LoraConfig(task_type=TaskType.SEQ_CLS, r=r)

    lora_model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=num_labels)
    lora_model = get_peft_model(lora_model, lora_config)
    lora_model.print_trainable_parameters()

    lora_model.to(device)

    trainer = Trainer(
        model=lora_model,
        args=args,
        train_dataset=dataset["train"],
        eval_dataset=dataset["validation"],
        processing_class=tokenizer,
        data_collator=data_collator,
        compute_metrics=compute_metrics,
    )
    train_and_pretty_print_results(trainer)
    print()

R =  2


Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


trainable params: 78,342 || all params: 109,565,196 || trainable%: 0.0715


No label_names provided for model class `PeftModelForSequenceClassification`. Since `PeftModel` hides base models input arguments, if label_names is not given, label_names can't be set automatically within `Trainer`. Note that empty label_names list will be used instead.


eval_loss: 0.2273
eval_Accuracy: 0.9180
eval_Recall: 0.8657
eval_Precision: 0.8792
eval_F1: 0.8717
eval_runtime: 3.3366
eval_samples_per_second: 599.4070
eval_steps_per_second: 9.5910
epoch: 10.0000
Training time: 595.34 sec
Trainable parameters: 0.08M
CUDA memory used: 2260.31MB

R =  4


Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


trainable params: 152,070 || all params: 109,638,924 || trainable%: 0.1387


No label_names provided for model class `PeftModelForSequenceClassification`. Since `PeftModel` hides base models input arguments, if label_names is not given, label_names can't be set automatically within `Trainer`. Note that empty label_names list will be used instead.


eval_loss: 0.2281
eval_Accuracy: 0.9220
eval_Recall: 0.8755
eval_Precision: 0.8801
eval_F1: 0.8764
eval_runtime: 3.3300
eval_samples_per_second: 600.5920
eval_steps_per_second: 9.6090
epoch: 10.0000
Training time: 596.71 sec
Trainable parameters: 0.15M
CUDA memory used: 2254.64MB

R =  8


Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


trainable params: 299,526 || all params: 109,786,380 || trainable%: 0.2728


No label_names provided for model class `PeftModelForSequenceClassification`. Since `PeftModel` hides base models input arguments, if label_names is not given, label_names can't be set automatically within `Trainer`. Note that empty label_names list will be used instead.


eval_loss: 0.2178
eval_Accuracy: 0.9215
eval_Recall: 0.8729
eval_Precision: 0.8871
eval_F1: 0.8780
eval_runtime: 3.3137
eval_samples_per_second: 603.5640
eval_steps_per_second: 9.6570
epoch: 10.0000
Training time: 599.53 sec
Trainable parameters: 0.30M
CUDA memory used: 2254.84MB

R =  16


Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


trainable params: 594,438 || all params: 110,081,292 || trainable%: 0.5400


No label_names provided for model class `PeftModelForSequenceClassification`. Since `PeftModel` hides base models input arguments, if label_names is not given, label_names can't be set automatically within `Trainer`. Note that empty label_names list will be used instead.


eval_loss: 0.2199
eval_Accuracy: 0.9220
eval_Recall: 0.8733
eval_Precision: 0.8934
eval_F1: 0.8793
eval_runtime: 3.3092
eval_samples_per_second: 604.3830
eval_steps_per_second: 9.6700
epoch: 10.0000
Training time: 600.56 sec
Trainable parameters: 0.59M
CUDA memory used: 2261.26MB

R =  32


Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


trainable params: 1,184,262 || all params: 110,671,116 || trainable%: 1.0701


No label_names provided for model class `PeftModelForSequenceClassification`. Since `PeftModel` hides base models input arguments, if label_names is not given, label_names can't be set automatically within `Trainer`. Note that empty label_names list will be used instead.


eval_loss: 0.2169
eval_Accuracy: 0.9245
eval_Recall: 0.8780
eval_Precision: 0.8919
eval_F1: 0.8833
eval_runtime: 3.3303
eval_samples_per_second: 600.5430
eval_steps_per_second: 9.6090
epoch: 10.0000
Training time: 597.98 sec
Trainable parameters: 1.18M
CUDA memory used: 2265.72MB



In [ ]:
data = [
    {
        "R": 2,
        "Trainable Params (M)": 0.08,
        "Trainable %": 0.0715,
        "Accuracy": 0.9180,
        "Recall": 0.8657,
        "Precision": 0.8792,
        "F1": 0.8717,
        "Training Time (s)": 595.34,
        "Eval Time (s)": 3.3366,
        "CUDA Memory (MB)": 2260.31,
    },
    {
        "R": 4,
        "Trainable Params (M)": 0.15,
        "Trainable %": 0.1387,
        "Accuracy": 0.9220,
        "Recall": 0.8755,
        "Precision": 0.8801,
        "F1": 0.8764,
        "Training Time (s)": 596.71,
        "Eval Time (s)": 3.3300,
        "CUDA Memory (MB)": 2254.64,
    },
    {
        "R": 8,
        "Trainable Params (M)": 0.30,
        "Trainable %": 0.2728,
        "Accuracy": 0.9215,
        "Recall": 0.8729,
        "Precision": 0.8871,
        "F1": 0.8780,
        "Training Time (s)": 599.53,
        "Eval Time (s)": 3.3137,
        "CUDA Memory (MB)": 2254.84,
    },
    {
        "R": 16,
        "Trainable Params (M)": 0.59,
        "Trainable %": 0.5400,
        "Accuracy": 0.9220,
        "Recall": 0.8733,
        "Precision": 0.8934,
        "F1": 0.8793,
        "Training Time (s)": 600.56,
        "Eval Time (s)": 3.3092,
        "CUDA Memory (MB)": 2261.26,
    },
    {
        "R": 32,
        "Trainable Params (M)": 1.18,
        "Trainable %": 1.0701,
        "Accuracy": 0.9245,
        "Recall": 0.8780,
        "Precision": 0.8919,
        "F1": 0.8833,
        "Training Time (s)": 597.98,
        "Eval Time (s)": 3.3303,
        "CUDA Memory (MB)": 2265.72,
    },
]

df_lora = pd.DataFrame(data)
df_lora = df_lora.style.highlight_max(subset=["Accuracy", "Recall", "Precision", "F1"], color="green")
df_lora

- качество постепенно растёт с увеличением r, но после r=16 — почти не меняется

- при r=32 достигается лучший F1 = 0.8833, но прирост незначителен по сравнению с r=16

- использование GPU памяти почти не растёт, что делает LoRA очень экономичным

- даже при r=32 обучается только 1.07% параметров модели (~1.18M из 110M)

- если важна эффективность, оптимальным можно считать r=8 или r=16

- r=2 — самый лёгкий, но теряет 1.1% F1 по сравнению с r=32

- в итоге я бы все равно выбрала как итоговый вариант r=32, так как его метрики лучше, хоть и незначительно

# 6. Итоговое сравнение

In [ ]:
data = {
    "method": ["no finetuning", "finetuning", "linear probing", "prefix tuning", "prompt tuning", "LoRA_r32"],
    "accuracy": [0.5775, 0.9235, 0.4645, 0.3820, 0.4750, 0.9245],
    "recall": [0.3039, 0.8891, 0.2388, 0.1901, 0.2989, 0.8780],
    "precision": [0.2751, 0.8951, 0.1532, 0.1284, 0.2231, 0.8919],
    "f1": [0.2425, 0.8823, 0.1858, 0.1400, 0.2508, 0.8833],
    "trainable_params_M": [109.49, 109.49, 0.00, 0.37, 0.02, 1.18],
    "cuda_mem_MB": [1334.25, 1334.25, 1774.06, 2256.85, 1804.14, 2265.72],
    "training_time_sec": [None, 1127.64, 315.08, 581.81, 780.66, 597.98],
}

result_df = pd.DataFrame(data)
result_df = (
    result_df.style.format(precision=4)
    .highlight_max(subset=["accuracy", "recall", "precision", "f1"], color="green")
    .highlight_min(subset=["trainable_params_M", "cuda_mem_MB", "training_time_sec"], color="green")
)
result_df

### коротко про результаты:
- лучшие метрики показал LoRA_r32 — чуть выше по всем показателям, чем полное fine-tuning, особенно по F1 (0.8833 против 0.8823). при этом обучается он почти в 2 раза быстрее (598 сек против 1128) и требует в 100 раз меньше параметров (1.18М vs 109.49М).

- fine-tuning тоже мощный, но дорогой и тяжёлый. много параметров, долго обучается.

- prompt tuning оказался лучше, чем linear и prefix tuning, но всё равно F1 у него всего 0.25 — слабо.

- prefix tuning и linear probing — худшие по всем метрикам. особенно precision: 0.128 и 0.153 соответственно.

- no finetuning удивительно дал F1 выше, чем prefix tuning и linear probing — то есть лучше вообще ничего не трогать, чем использовать эти два. либо досконально настраивать параметры, так как я использовала одинаковые для всех.

- в целом, LoRA показала себя лучше всех — почти как фулл fine-tuning, но сильно легче в ресурсах.